src/pool/srv_pool.c

```diff
@@ -2385,24 +2390,34 @@ pool_svc_step_up_cb(struct ds_rsvc *rsvc)
 		D_FREE(map_buf);
 	if (prop != NULL)
 		daos_prop_free(prop);
-	if (rc < 0)
-		ds_pool_failed_add(svc->ps_uuid, rc);
-	else if (rc == 0)
-		ds_pool_failed_remove(svc->ps_uuid);
+	if (svc->ps_error != 0) {
+		/*
+		 * Step up with the error anyway, so that RPCs to the PS
+		 * receive an error instead of timeouts.
+		 */
+		DS_POOL_NOTE_PRINT(DF_UUID": rank %u became pool service leader "DF_U64
+				   " with error: "DF_RC"\n", DP_UUID(svc->ps_uuid), rank,
+				   svc->ps_rsvc.s_term, svc->ps_error);
+		rc = 0;
+	}
 	return rc;
 }
 
 static void
 pool_svc_step_down_cb(struct ds_rsvc *rsvc)
 {
-	struct pool_svc	       *svc = pool_svc_obj(rsvc);
-	d_rank_t		rank = dss_self_rank();
+	struct pool_svc *svc  = pool_svc_obj(rsvc);
+	d_rank_t         rank = dss_self_rank();
 
-	pool_svc_step_down_metrics(svc);
-	fini_events(svc);
-	sched_cancel_and_wait(&svc->ps_reconf_sched);
-	sched_cancel_and_wait(&svc->ps_rfcheck_sched);
-	ds_cont_svc_step_down(svc->ps_cont_svc);
+	if (svc->ps_error == 0) {
+		pool_svc_step_down_metrics(svc);
+		fini_events(svc);
+		sched_cancel_and_wait(&svc->ps_reconf_sched);
+		sched_cancel_and_wait(&svc->ps_rfcheck_sched);
+		ds_cont_svc_step_down(svc->ps_cont_svc);
+	} else {
+		svc->ps_error = 0;
```

Comment:
Not have the full understanding, but it seems that some error messages could be hidden here ?
Or maybe we are in the same situation as the comment of line 2530 ?

Reply:
My apologies as my remark was confusing and indeed I was not referring to the Raft leadership.
Just thinking that this error was managed by upper layer and thus should not be logged.

Reply:
@knard38, no problem at all. This PR introduces a new up-but-not-working-due-to-error mode to PS, so that we can avoid using `ds_pool_failed_add`. So the error in `ps_error` is part of this special mode. I think it's okay to log the error twice: once during stepping up and once during stepping down. For instance,

    rank R became pool service leader 3 with error: <some error>
    ...
    rank R no longer pool service leader 3 with error: <some error>